File: backend/candidate-profile-service/src/main/java/com/jobportal/candidateprofile/controllers/CandidateExperienceController.java
package com.jobportal.candidateprofile.controllers;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.jobportal.candidateprofile.dto.CandidateExperienceDto;
import com.jobportal.candidateprofile.service.CandidateExperienceService;

@RestController
@RequestMapping("/api/experiences")
@CrossOrigin("*")
public class CandidateExperienceController {

    @Autowired
    private CandidateExperienceService service;

    @PostMapping
    public CandidateExperienceDto addExperience(@RequestBody CandidateExperienceDto dto) {
        return service.addExperience(dto);
    }

    @PutMapping("/{expId}")
    public CandidateExperienceDto updateExperience(@PathVariable Integer expId, @RequestBody CandidateExperienceDto dto) {
        return service.updateExperience(expId, dto);
    }

    @DeleteMapping("/{expId}")
    public String deleteExperience(@PathVariable Integer expId) {
        service.deleteExperience(expId);
        return "Experience deleted successfully";
    }

    @GetMapping("/{expId}")
    public CandidateExperienceDto getExperienceById(@PathVariable Integer expId) {
        return service.getExperienceById(expId);
    }

    @GetMapping("/candidate/{cid}")
    public List<CandidateExperienceDto> getExperiencesByCandidate(@PathVariable Integer cid) {
        return service.getExperiencesByCandidate(cid);
    }

    @GetMapping
    public List<CandidateExperienceDto> getAllExperiences() {
        return service.getAllExperiences();
    }
}

File: backend-dotnet/Services/JobPortal.CandidateService/Controllers/CandidateExperienceController.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using JobPortal.CandidateService.DTOs;
using JobPortal.CandidateService.Services;

namespace JobPortal.CandidateService.Controllers;

[ApiController]
[Route("api/experiences")]
public class CandidateExperienceController : ControllerBase
{
    private readonly CandidateExperienceService _service;

    public CandidateExperienceController(CandidateExperienceService service)
    {
        _service = service;
    }

    [HttpPost]
    public async Task<IActionResult> AddExperience([FromBody] CandidateExperienceDto dto)
    {
        var result = await _service.AddExperienceAsync(dto);
        return Ok(result);
    }

    [HttpPut("{expId}")]
    public async Task<IActionResult> UpdateExperience(int expId, [FromBody] CandidateExperienceDto dto)
    {
        try
        {
            var result = await _service.UpdateExperienceAsync(expId, dto);
            return Ok(result);
        }
        catch (KeyNotFoundException)
        {
            return NotFound("Experience not found");
        }
    }

    [HttpDelete("{expId}")]
    public async Task<IActionResult> DeleteExperience(int expId)
    {
        try
        {
            await _service.DeleteExperienceAsync(expId);
            return Ok("Experience deleted successfully");
        }
        catch (KeyNotFoundException)
        {
            return NotFound("Experience not found");
        }
    }

    [HttpGet("{expId}")]
    public async Task<IActionResult> GetExperienceById(int expId)
    {
        try
        {
            var result = await _service.GetExperienceByIdAsync(expId);
            return Ok(result);
        }
        catch (KeyNotFoundException)
        {
            return NotFound("Experience not found");
        }
    }

    [HttpGet("candidate/{cid}")]
    public async Task<IActionResult> GetExperiencesByCandidate(int cid)
    {
        var list = await _service.GetExperiencesByCandidateAsync(cid);
        return Ok(list);
    }

    [HttpGet]
    public async Task<IActionResult> GetAllExperiences()
    {
        var list = await _service.GetAllExperiencesAsync();
        return Ok(list);
    }
}
